cpp/src/strings/split/split_re.cu

```diff
@@ -0,0 +1,374 @@
+ */
+
+#include <strings/count_matches.hpp>
+#include <strings/regex/regex.cuh>
+#include <strings/utilities.hpp>
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/get_value.cuh>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/strings/detail/strings_column_factories.cuh>
+#include <cudf/strings/split/split_re.hpp>
+#include <cudf/strings/string_view.cuh>
+#include <cudf/strings/strings_column_view.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+
+#include <thrust/for_each.h>
+#include <thrust/transform_scan.h>
+
+namespace cudf {
+namespace strings {
+namespace detail {
+namespace {
+
+using string_index_pair = thrust::pair<const char*, size_type>;
+
+enum class split_direction {
+  FORWARD,  ///< for split logic
+  BACKWARD  ///< for rsplit logic
+};
+
+/**
+ * @brief Identify the tokens from the `idx'th` string element of `d_strings`.
+ *
+ * Each string's tokens are stored in the `d_tokens` vector.
+ * The `d_token_offsets` specifies the output position within `d_tokens`
+ * for each string.
+ */
+template <int stack_size>
+struct token_reader_fn {
+  column_device_view const d_strings;
+  reprog_device prog;
+  split_direction const direction;
+  offset_type const* d_token_offsets;
+  string_index_pair* d_tokens;
+
+  __device__ void operator()(size_type idx)
+  {
+    if (d_strings.is_null(idx)) { return; }
+    auto const d_str = d_strings.element<string_view>(idx);
+
+    auto const token_offset = d_token_offsets[idx];
+    auto const token_count  = d_token_offsets[idx + 1] - token_offset;
+    auto const d_result     = d_tokens + token_offset;  // store tokens here
+
+    size_type token_idx = 0;
+    size_type begin     = 0;  // characters
+    size_type end       = d_str.length();
+    size_type last_pos  = 0;  // bytes
+    while (prog.find<stack_size>(idx, d_str, begin, end) > 0) {
+      // get the token (characters just before this match)
+      auto const token =
+        string_index_pair{d_str.data() + last_pos, d_str.byte_offset(begin) - last_pos};
+      // store it if we have space
+      if (token_idx < token_count - 1) {
+        d_result[token_idx++] = token;
+      } else {
+        if (direction == split_direction::FORWARD) { break; }  // we are done
+        for (auto l = 0; l < token_idx - 1; ++l) {
+          d_result[l] = d_result[l + 1];  // shift left
+        }
+        d_result[token_idx - 1] = token;
+      }
+      // setup for next match
+      last_pos = d_str.byte_offset(end);
+      begin    = end + (begin == end);
+      end      = d_str.length();
+    }
+
+    // set the last token to the remainder of the string
+    d_result[token_idx] = string_index_pair{d_str.data() + last_pos, d_str.size_bytes() - last_pos};
+
+    if (direction == split_direction::BACKWARD) {
+      // update first entry -- this happens when max-tokens is hit before the end of the string
```

Comment:
```suggestion
      // update first entry -- this happens when maxsplit is hit before the end of the string
```

Reply:
The `max-tokens` is actually more correct here. I feel it is more clear what is actually happening.

Reply:
Sounds good. I was 50/50 on this suggestion, just thought it was worth proposing.